File: dashboard/src/stores/reviews.ts
import { defineStore } from 'pinia'
import { ref } from 'vue'

export interface ReviewSummary {
  id: number
  delivery_id: string
  owner: string
  repo: string
  pr_number: number
  head_sha: string
  model: string | null
  input_tokens: number
  output_tokens: number
  total_findings: number
  created_at: string | null
}

export interface Finding {
  axis: string
  file: string
  line: number | null
  issue: string
  suggestion: string
  source: string
}

export interface ReviewDetail extends Omit<ReviewSummary, 'total_findings'> {
  findings: Finding[]
}

export interface Stats {
  total_reviews: number
  total_findings: number
  total_input_tokens: number
  total_output_tokens: number
}

export const useReviewStore = defineStore('reviews', () => {
  const reviews = ref<ReviewSummary[]>([])
  const current = ref<ReviewDetail | null>(null)
  const stats = ref<Stats>({ total_reviews: 0, total_findings: 0, total_input_tokens: 0, total_output_tokens: 0 })
  const loading = ref(false)

  async function fetchReviews(offset = 0, limit = 20) {
    loading.value = true
    try {
      const res = await fetch(`/api/reviews?offset=${offset}&limit=${limit}`)
      reviews.value = await res.json()
    } finally {
      loading.value = false
    }
  }

  async function fetchReview(id: number) {
    loading.value = true
    try {
      const res = await fetch(`/api/reviews/${id}`)
      if (res.ok) {
        current.value = await res.json()
      }
    } finally {
      loading.value = false
    }
  }

  async function fetchStats() {
    const res = await fetch('/api/stats')
    stats.value = await res.json()
  }

  return { reviews, current, stats, loading, fetchReviews, fetchReview, fetchStats }
})

File: backend/src/pairo/api/reviews.py
from typing import Any

from fastapi import APIRouter, Response

from pairo.infrastructure.persistence.decision_repo import (
    SqlDecisionRepository,
)
from pairo.infrastructure.persistence.engine import get_session
from pairo.infrastructure.persistence.repository import SqlReviewRepository

router = APIRouter(prefix="/api")


def _repo() -> SqlReviewRepository:
    return SqlReviewRepository(get_session())


@router.get("/reviews")
async def list_reviews(offset: int = 0, limit: int = 20) -> list[dict[str, object]]:
    repo = _repo()
    reviews = await repo.list_reviews(offset=offset, limit=limit)
    return [
        {
            "id": r.id,
            "delivery_id": r.delivery_id,
            "owner": r.owner,
            "repo": r.repo,
            "pr_number": r.pr_number,
            "head_sha": r.head_sha,
            "model": r.model,
            "input_tokens": r.input_tokens,
            "output_tokens": r.output_tokens,
            "total_findings": r.total,
            "created_at": str(r.created_at) if r.created_at else None,
        }
        for r in reviews
    ]


@router.get("/reviews/{review_id}", response_model=None)
async def get_review(review_id: int) -> dict[str, object] | Response:
    repo = _repo()
    r = await repo.get(review_id)
    if r is None:
        return Response(
            content='{"detail":"Not found"}',
            status_code=404,
            media_type="application/json",
        )
    return {
        "id": r.id,
        "delivery_id": r.delivery_id,
        "owner": r.owner,
        "repo": r.repo,
        "pr_number": r.pr_number,
        "head_sha": r.head_sha,
        "model": r.model,
        "input_tokens": r.input_tokens,
        "output_tokens": r.output_tokens,
        "created_at": str(r.created_at) if r.created_at else None,
        "findings": [
            {
                "axis": f.axis.value,
                "file": f.file,
                "line": f.line,
                "issue": f.issue,
                "suggestion": f.suggestion,
                "source": f.source.value,
                "category": f.category,
            }
            for f in r.findings
        ],
    }


@router.get("/stats")
async def stats() -> dict[str, int]:
    repo = _repo()
    return await repo.stats()


@router.get("/stats/memory")
async def memory_stats() -> dict[str, Any]:
    session = get_session()
    decision_repo = SqlDecisionRepository(session)
    return await decision_repo.memory_stats()
